### ISSUE TITLE
Load Order resets to alphabetical order when closing MO2 with another program open

### ISSUE THREAD

**Issue by VivaNewVegas** [ModOrganizer2/modorganizer#1534]
## The problem:
Stumbled upon this when I accidentally closed MO2 instead of the GECK. If you close MO2 (with MO2 *still locked*) while you still have a program you ran through MO2 open (be it the game itself, xEdit, GECK), then your load order will be set to alphabetical order upon relaunching MO2. No prompt shows up saying "Waiting for program to close before closing MO2" if you do not unlock MO2 first, but if you do unlock it then the prompt does show up and the load order stays untouched. This has only been tested on a New Vegas instance on v2.4.2

## To Reproduce:
1. Run any program through Mod Organizer 2 (tested with New Vegas, GECK, and xEdit)
2. Alt-tab from the program to get back to MO2
3. Close MO2 using the red x in the top right corner WITHOUT unlocking the program first 
4. This will instantly close MO2 without any prompts, but the program you launched through MO2 will stay open
5. Once you close said program and relaunch MO2, the load order will be reset to alphabetical order (in my testing it only affected the load order and not the mod order)

## Environment:
* Version 2.4.2 released on Nexus
* Portable instance located on separate drive from Windows installation, exceptions have been made in Windows Defender
* Windows 10 Pro 20H2
* Ryzen 7 1700, 16gb RAM, GTX 1060 6GB

## Details:
Proof of Defender Exclusions: https://imgur.com/a/0dzPipv

### USVFS:
USVFS Log: https://gist.github.com/VivaNewVegas/e3b9c02cfb245b3fe1483a2d836974e8

### MO Interface:
Interface Log: https://gist.github.com/VivaNewVegas/ae88cc1aa802f6658eb7f9a5a3db1199

**Comment by isanae** [comment]
I can't reproduce this here with a quick test. MO does close without a prompt, but my load order is not reset when I restart it. I agree the lack of a prompt is a bit meh, MO handles that case the same way as for the "waiting on an application to close before exiting" warning, it just exits. It's the same code iirc, the text is just different.

**Comment by VivaNewVegas** [comment]
any other info you need me to provide? You can also just close this if you're confident its not an issue

**Comment by isanae** [comment]
I'm just saying I can't reproduce it with my setup, I don't doubt it's happening to you. We can leave this open in case someone has more info.

**Comment by ucwr** [comment]
In my case, it always happens when MO2 is closed accordingly scenario described by VivaNewVegas. I encountered this issue countless times for the last year. I don't know why you cannot reproduce it, my wild guess is it has soomething to do with different Windows versions (mine is 21H1).

It's a shame that this bug was not fixed yet. Because it's a very nasty one:
- it's quite sneaky: I usually notice that something went wrong after many (lost) hours of gameplay;
- an incorrect load order almost certainly brokes quests' flow (among other things) and hence saves;
- to find the last "unbroken" save I have to inspect my cosaves (.nvse for F:NV, .fo4se for FO4, etc) in an external editor.

---
As a workaround, now I keep my entire load order locked.  It 100% helps but there is a downside (probably another issue): after deinstalling a mod having a locked plugin it becomes a "zombie" (it stays in the "lockedorder.txt" file forever and messes with the load order).

**So, this is what I usually do:**
1. close the MO2 application and manually remove the "lockedorder.txt" (not "loadorder.txt"!) file from the active profile;
2. start the MO2 and check the load order and correct it, if necessary;
3. lock it (select all the plugins (LMB click on any one, then Ctrl-A), RMB click, "Lock load order");
4. make a backup.

**Comment by LostDragonist** [comment]
@ucwr Can you try reproducing the issue with a very small number of mods (like... 2 or 3)? I'm also unable to reproduce at all using GOG, Steam, and 3 different versions of Windows.  I suspect this is an issue with a very specific mod that us devs probably don't use.

**Comment by ucwr** [comment]
Yes. Just now I tested with only two and then three .esm plugins. Each time the load order was reset.

**Comment by ucwr** [comment]
I use a portable version if this matters.

**Comment by LostDragonist** [comment]
Okay, what are the plugins?

**Comment by ucwr** [comment]
I chose random ones: Lonesomeroad.esm, HonestHearst.esm and GunRunnerArsenal.esm. But I can reproduce the issue with any combination of plugins on my side.

**Comment by LostDragonist** [comment]
Alright, managed to reproduce. Guess I did something different this time? IDK.   

Working log where GECK is closed before MO2:
```
[2022-01-18 21:29:01.942 D] waiting for completion on Geck.exe (14620), strong interest
[2022-01-18 21:29:13.830 D] process 14620 completed
[2022-01-18 21:29:13.831 D] nothing to wait for
[2022-01-18 21:29:13.833 D] process runner: refreshing because the process completed
[2022-01-18 21:29:13.833 D] removing loadorder.txt
[2022-01-18 21:29:13.833 D] refreshing structure
[2022-01-18 21:29:13.833 D] timing: OrganizerCore::refreshESPList() 0 ms
[2022-01-18 21:29:13.837 D] refresher: using 10 threads
[2022-01-18 21:29:13.839 D] refresher saw 355 files
[2022-01-18 21:29:13.839 D] timing: DirectoryRefresher::refresh() 4 ms
[2022-01-18 21:29:13.873 D] directory refreshed, finishing up
[2022-01-18 21:29:13.873 D] clearing caches
[2022-01-18 21:29:13.873 D] running 2 post refresh tasks
[2022-01-18 21:29:13.873 D] structure deleter thread start
[2022-01-18 21:29:13.874 D] structure deleter thread done
[2022-01-18 21:29:13.878 D] timing: PluginList::refresh() 5 ms
[2022-01-18 21:29:13.878 D] timing: OrganizerCore::refreshESPList() 5 ms
[2022-01-18 21:29:13.924 D] setting file times on esps
[2022-01-18 21:29:13.924 D] refreshing lists
[2022-01-18 21:29:13.929 D] timing: PluginList::refresh() 3 ms
[2022-01-18 21:29:13.929 D] timing: OrganizerCore::refreshESPList() 4 ms
[2022-01-18 21:29:13.930 D] timing: OrganizerCore::refreshBSAList() 0 ms
[2022-01-18 21:29:13.930 D] refresh done
[2022-01-18 21:29:13.930 D] timing: OrganizerCore::directory_refreshed() 56 ms
```

Broken log where MO2 is closed before GECK:

```
[2022-01-18 21:30:08.791 D] waiting for completion on Geck.exe (9144), strong interest
[2022-01-18 21:30:19.740 D] waiting for 9144 interrupted
[2022-01-18 21:30:19.741 D] process runner: refreshing because the ui was force unlocked
[2022-01-18 21:30:19.741 D] removing loadorder.txt
[2022-01-18 21:30:19.741 D] refreshing structure
[2022-01-18 21:30:19.741 D] timing: OrganizerCore::refreshESPList() 0 ms
[2022-01-18 21:30:19.745 D] refresher: using 10 threads
[2022-01-18 21:30:19.751 D] nexus: connection cancelled
[2022-01-18 21:30:19.756 D] refresher saw 355 files
[2022-01-18 21:30:19.756 D] timing: DirectoryRefresher::refresh() 14 ms
[2022-01-18 21:30:19.764 W] not saving lists during directory update
[2022-01-18 21:30:19.842 D] nexus: connection cancelled
[2022-01-18 21:30:19.843 D] save C:/Users/USERNAME/AppData/Local/ModOrganizer/New Vegas/webcache/nexus_cookies.dat
```

So... yeah, MO2 is just yeeting the loadorder.txt.  That's nice.  Also explains why this disproportionally affects New Vegas and Oblivion as newer games don't use that file.

**Comment by LostDragonist** [comment]
I guess the main questions here are going to be:
1. Why is there so much crap between loadorder.txt being deleted and loadorder.txt being created again.
2. If we're not going to be writing loadorder.txt, why is it being deleted?

**Comment by Al12rs** [comment]
If I remember right for older games the load order is decided by plugin time stamps. So if you run loot for example it has to change the timestamps of the plugins, then Mo2 needs to rebuild its internal load order based on those timestamps loadorder.txt.

The reason why it's deleting loadorder.txt is because MO2 is delegating the definition of the plugin order to the game plugin (so it can behave differently for games that use timestamps or not for example) and to force the game plugin to create the order based on the timestamps instead of the internal loadorder.txt it deletes it. I guess it was the easiest way to do it or something. 

I don't know why it won't write the new order correctly afterwards though. I guess one possibility is that Mo2 isn't rewriting the plugin timestamps correctly itself, so if no one else does it then they probably all have the same timestamp or something? I don't know. 

So either the plugins timestamps are messed up so deleting loadorder.txt simply reads the messed up timestamps, or the timestamps are correct and Mo2 messed up when writing back that loadorder.txt for some reason.

**Comment by LostDragonist** [comment]
> [2022-01-18 21:30:19.764 W] not saving lists during directory update

This is the key part of the log.  MO2 is in the middle of updating the file tree so it thinks it can't write out loadorder.txt et al. But loadorder.txt has already been deleted so ...... gg?

**Comment by Al12rs** [comment]
Yeah, but logically it shouldn't matter because it's just going to read the load order from the timestamps again, so if that one is correct then it should work anyways in theory.

**Comment by LostDragonist** [comment]
The problem is that MO2 is closed.  There's nothing after that.  It's already in the process of shutting down when load_order.txt is deleted and the directory is being refreshed. So there's nothing left to write load_order.txt because the program is dead.

**Comment by Al12rs** [comment]
Yeah but if loadorder.txt is missing it shouldn't be a problem as long as the timestamps are correct, as if it is missing MO2 will recreate it next time it starts by using the timestamps.

**Comment by LostDragonist** [comment]
Then I guess that code is broken or not doing what it's supposed to.  Cause it certainly isn't doing that.

**Comment by Al12rs** [comment]
My conclusion is that the problem is with the timestamps. Mo2 is messing up writing them. Either because it is missing loadorder.txt or perhaps for some other reason and the timestamps are simply wrong whey they are being read.

**Comment by LostDragonist** [comment]
There is no problem with the timestamps. They remain constant through the entire process.

**Comment by LostDragonist** [comment]
![image](https://user-images.githubusercontent.com/7862539/150034083-910149c9-c1cd-4254-88df-9639750d0095.png)

Here you can see that MO hasn't even attempted to read or write the timestamps from the files.  It's just... doing nothing. /shrug.

**Comment by Al12rs** [comment]
If you change the load order do the timestamps change?

**Comment by LostDragonist** [comment]
If I change the load order inside MO2, or otherwise nudge the plugin list to update, sure it write's out the correct timestamps.

**Comment by Al12rs** [comment]
So the timestamps remain correct through the whole thing? How does this get messed up for the game then? They have to change to be incorrect at some point or the issue wouldn't be there

**Comment by LostDragonist** [comment]
I'd imagine it's not actually an issue until the user sees the load "has changed" and then goes to fix it at which point MO2 will gladly write out the timestamps to match the plugin list.  There's something wrong with the logic being used to determine whether to use loadorder.txt or plugins.txt to set the priority of the plugin list.  Just trying to run through it to see what's happening.

**Comment by LostDragonist** [comment]
Okay, make that... huh. At some point during startup, loadorder.txt is created before we even check that it exists...?  So it thinks loadorder.txt is valid because it existed before so it just doesn't go on to the timestamps.  That's......... where the hell is that happening?

**Comment by LostDragonist** [comment]
That's happening in GameFalloutNV::initializeProfile.  It sees no file exists so it creates one cause it's supposed to exist.  That's..... fun.

**Comment by LostDragonist** [comment]
Yeeep. Preventing initializeProfile from writing that blank file fixes it.  Judging by the other code surrounding loadorder.txt (and Al's comments), that's probably gonna be the fix? Though gonna sleep on it.

**Comment by Al12rs** [comment]
Yeah that sounds like it could just fix the problem, though the whole system isn't the best. 

What happens to the timestamps on MO2 startup if the loadorder.txt is missing? Do they get messed up or are they left to be incongruent with how Mo2 shows them?

**Comment by LostDragonist** [comment]
With the current code, https://github.com/ModOrganizer2/modorganizer/issues/1534#issuecomment-1015920024 happens. MO2 doesn't touch the timestamps until it has a reason to touch the timestamps. This is going to be some sort of user action that ends up requiring the loadorder.txt to be regenerated.  

In other words, in this scenario, MO2 is straight up lying about what the load order is.

**Comment by Al12rs** [comment]
Yeah I really don't like that, there should be some kind of sync between the two. It would have wrecked the load order of our users with the current code if it did that but at least we would have noticed the bug.

**Comment by ucwr** [comment]
Regardless of the timestamps issue, I have to say, the current behavior, when "loadorder.txt" is deleted even before its replacement was created seems to be a little bit dangerous., don't you think? I mean, shouldn't there be something like this: (1) create "loadorder.tmp", (2) recreeate "loadorder.txt" with new data, (3) delete "loadorder.tmp"? Then, if something goes wrong, "loadorder.tmp"'s presence will be a hint for MO when it starts. I am not judging though: there could be some reason for such decision at the time.

Besides, although this is probably not going to happen often, but MO still can be closed by Windows, if for some reason it decided to restart itself AND an application launched by MO finished/terminated first.
By the way, when saveCurrentLists() refuses to save the file, there is no second try, am I right? When the "loadorder.txt" was already erased?

Another question. Why MO doesn't use "plugins.txt" as the source of the info during "loadorderless" start, while normally it effectively mirrors the file content to the "loadorder.txt" anyway?

Also, I noticed that the load order is always correct (or, at least, it was so on my side), if MO was unlocked before exiting (when there appears a modal dialog with "Exit Now", "Cancel" buttons).

**Comment by LostDragonist** [comment]
> Regardless of the timestamps issue, I have to say, the current behavior, when "loadorder.txt" is deleted even before its replacement was created seems to be a little bit dangerous., don't you think? I mean, shouldn't there be something like this: (1) create "loadorder.tmp", (2) recreeate "loadorder.txt" with new data, (3) delete "loadorder.tmp"? Then, if something goes wrong, "loadorder.tmp"'s presence will be a hint for MO when it starts. I am not judging though: there could be some reason for such decision at the time.

Yes, it absolutely is dangerous.  /shrug.  You'd have to ask the original author why stuff is the way it is.

> By the way, when saveCurrentLists() refuses to save the file, there is no second try, am I right? When the "loadorder.txt" was already erased?

There is a method in place for writing loadorder.txt after this happens but the way it was implemented was using a Qt event.  However Qt events aren't processed when the application is shutting down so this isn't effective in this case. This seems like an oversight and easy enough to do.

> Another question. Why MO doesn't use "plugins.txt" as the source of the info during "loadorderless" start, while normally it effectively mirrors the file content to the "loadorder.txt" anyway?

For the games in question, plugins.txt only contains any active plugins. Inactive plugins would thus lose their priorities were it not for loadorder.txt.  That might be a little better in such a failure case like this but we can also just fix the failure case.

> Also, I noticed that the load order is always correct (or, at least, it was so on my side), if MO was unlocked before exiting (when there appears a modal dialog with "Exit Now", "Cancel" buttons).

Yes. That's because this forces the directory refresh to happen while MO is not in the process of closing.

**Comment by ucwr** [comment]
> You'd have to ask the original author why stuff is the way it is.

Yes, I understand and know that such things "just happens" sometimes.

> <...> That might be a little better in such a failure case like this but we can also just fix the failure case.

Agreed.
Thanks for thorough replies!

**Comment by Al12rs** [comment]
> Regardless of the timestamps issue, I have to say, the current behavior, when "loadorder.txt" is deleted even before its replacement was created seems to be a little bit dangerous., don't you think? I mean, shouldn't there be something like this: (1) create "loadorder.tmp", (2) recreeate "loadorder.txt" with new data, (3) delete "loadorder.tmp"? Then, if something goes wrong, "loadorder.tmp"'s presence will be a hint for MO when it starts. I am not judging though: there could be some reason for such decision at the time.

The reason is that the true order is decided by timestamps, not loadorder.txt. In theory loadorder.txt is just a cache to avoid having to read the timestamps time and time again. So deleting loadorder.txt is like refreshing the cache.

The actual problem is when loadorder.txt is present but the contents aren't matching timestamps, like when there is an empty one on startup in the case of this bug.

**Comment by ucwr** [comment]
> The reason is that the true order is decided by timestamps, not loadorder.txt. In theory loadorder.txt is just a cache to avoid having to read the timestamps time and time again. So deleting loadorder.txt is like refreshing the cache.

> The actual problem is when loadorder.txt is present but the contents aren't matching timestamps, like when there is an empty one on startup in the case of this bug.

But not only empty, the absent file also ruins the load order! I tested it right now.

I did not notice the issue until early July, when I updated my Windows from v.1909. But I didn't use MO2 (and did not touch F:NV) for almost a year, so when I updated Windows, FNV and MO2, I got a nightmarish week, trying to make FNV to work without CTD every minute. This is when I noticed the issue (yes, I saw the bug report, but it was very clean, what could I add?).
So, maybe MO2 did work as you described (and as it supposed to), because I really did not notice anything strange before the said updates.
Thanks!

**Comment by LostDragonist** [comment]
> But not only empty, the absent file also ruins the load order! I tested it right now.

I mean... yes... that's why we're here in the first place and need to fix the bug.

**Comment by Al12rs** [comment]
> But not only empty, the absent file also ruins the load order! I tested it right now.

From what I can tell the problem is that MO2 is just looking at your loadorder.txt and assuming that timestamps are reflecting that while your timestamps are probably rubbish.  If you delete the loadorder.txt it reads the rubbish from the timestamps. 

Mo2 only writes the timestamps when you try to change the order apparently.

**Comment by ucwr** [comment]
> From what I can tell the problem is that MO2 is just looking at your loadorder.txt and assuming that timestamps are reflecting that while your timestamps are probably rubbish. If you delete the loadorder.txt it reads the rubbish from the timestamps.
> 
> Mo2 only writes the timestamps when you try to change the order apparently.

You know what? You are right! I checked the plugins and when I use to sort them by date/time in a file manager (it an external one, but I opened it through MO2), they also are ordered in strict alphanumeric order. WT...?!

**Comment by ucwr** [comment]
So. I knew that GECK loves to set 2000 year to an edited file. But this is too much.

<details>
  <summary>timestamps (I cut out the extra lines)</summary>

````
03/31/2000 08:00     <- Zeta.esm
03/30/2000 08:00     <- YUPTTW.esm
03/29/2000 08:00     < - WMVM.esm
----- other 25 lines --
03/03/2000 07:00
03/02/2000 07:00
03/01/2000 07:00
02/29/2000 07:00
02/28/2000 07:00
----- other 26 lines --
02/01/2000 07:00
01/31/2000 07:00
01/30/2000 07:00
----- other 26 lines  --
01/03/2000 07:00     <- A Trail of Crumbs.esm
01/02/2000 07:00     <- _3rd Camera Position.esp
01/01/2000 07:00     <- FalloutNV.esm (the only exception)
````
</details>

What program could do that? GECK? FNV's plugin? MO?

**Comment by LostDragonist** [comment]
That's MO2 doing that.  I know the problem.  I can fix it.  I just apparently can't explain it well enough to make the discussion go away.  

So I'm going to ask to just leave it be until I can get some time and motivation to fix it.

**Comment by ucwr** [comment]
Got it! Thanks! :)

**Comment by LostDragonist** [comment]
Should be fixed in 2.4.5alpha1.  Links available below.

Archive: <https://mega.nz/file/gqxCRT4R#sMBGujL_kMUV_TEDZ9qt7YSPtBKuBC7Q_128VjPwqJI>
Installer: <https://mega.nz/file/A3xUzLJS#0y0HD1ptG0hH5f6AsHGfdFd-hwbBleTefshpMgnJfFs>
PDBs (for devs): <https://mega.nz/file/Ymhw3bYA#_Q6Z99vjRHU0AswxJyB9uYCxoIxISvEKY8-2vEbqOp4>
Source: <https://mega.nz/file/ozgmgb4a#7cAmp2SbRSlkLGV9v-sI3x4uej0yUn2328YleITgSZU>

**Comment by ucwr** [comment]
> Should be fixed in 2.4.5alpha1. Links available below.

Seems works very well. Thank you so much!

One thing though. I've noticed that MO now prevents CaravanPack.esm and ClassicPack.esm from reordering: it keep them (along with FalloutNV.esm and right after it) in  sort of "unmanageable zone". Although this is not game breaking behavior, I suppose, but still?

**Comment by Al12rs** [comment]
From the changelog:
> Improvements:
* Plugins in all Bethesda games will more rigorously follow plugin rules. Masters will be up top, normal plugins will be down below. Plugins must be loaded after their masters. These rules should be enforced all the time now instead of just when changing something in the plugin list.

Does the behaviour make sense considering that now mo2 enforces plugins to be under their masters?

**Comment by ucwr** [comment]
> Does the behaviour make sense considering that now mo2 enforces plugins to be under their masters?

Of course, this is the correct behavior. But how this make those two esm pluging very special? Perhaps I was not enough clear in my message.

The correct order for F:NV is:
```
FalloutNV.esm (and exe)

DeadMoney.esm
HonestHearts.esm
OldWorldBlues.esm
LonesomeRoad.esm

GunRunnersArsenal.esm
ClassicPack.esm
MercenaryPack.esm
TribalPack.esm
CaravaPack.esm
```

(I think the load order of the files from the last group is not so important, since those plugins are injected ones. But TBH I am not entirely sure.)

But now MO locks the Caravan and Classic packs right after FalloutNV.
![loadorder](https://user-images.githubusercontent.com/3603562/150697229-c89dacdd-f84d-4409-a4b0-f29e49e778d2.png)

This is what I meant.
Or maybe there is something wrong in my settings?

**Comment by Al12rs** [comment]
Ah that is the other part of the changelog:
> * Fallout NV now recognizes .NAM files and will set DLC as force-loaded when present.

So I guess you only have the nam files for those two dlcs?

**Comment by ucwr** [comment]
Nope, I didn't know this, thanks, but I've checked and all of them are in the target directory. (Although I made a mod that let me to toggle them)

**Comment by Al12rs** [comment]
Can you confirm that the only .nam files that are in the game data folder (not mods) are CaravanPack.nam and ClassicPack.nam ?

**Comment by ucwr** [comment]
I found the reason.  I launched another MO with only F:NV and saw the same. So I checked the game's root directory and those two plugins were there! I removed them from there and now locks gone!
The only question I have is should the MO ignore NAMs from an archive that places them into the virtual root folder? If this is an intended and documented behavior, just ignore my question as utterly stupid and have my apologies.

**Comment by Al12rs** [comment]
No, it's not as intended. I asked to make sure that was what was going on. For now it's just checking the game data folder. It will need to also query the internal directory structure, which will require some checks to make sure is even ready to be searched.

**Comment by ucwr** [comment]
> No, it's not as intended. I asked to make sure that was what was going on.

Ah, did not see your question until the page's been reloaded, but answered them anyway, I think.

> For now it's just checking the game data folder. It will need to also query the internal directory structure, which will require some checks to make sure is even ready to be searched.
>
>Either that or some changes to how the order is enforced between force enabled DLCs and the other ones.

I see. I believe it's a very good compromise.

Well, I am not the original author, but at least on my side the issue was completely resolved. 
Thank you very, very much! I really appreciate your help, Al12rs, and your fix,  LostDragonist!

**Comment by LostDragonist** [comment]
Here's 2.4.5alpha2 which removes that .nam behavior. 
 https://mega.nz/file/1zRHiYgR#fVpQWYM_dHWSz2I-9F9SnnRvFussVMzDJ1mxZsHe3RM

This issue is now considered fixed so I'll be closing it.  If you happen to run into any other problems, please create a new issue or drop by the discord.
